### PR TITLE
Storage Distribution Algo support for AWS.

### PR DESCRIPTION
**Pull request by adityadani**
The changes are divided into 3 commits

1st commit - New yaml file for AWS. Changes to existing azure yaml file.
```
- The aws yaml is generated and not hand written.
- Introduced min and max iops in a row. Modified azure yaml to include this new change.
```

2nd commit - Code to generate AWS yaml

3rd commit - Support for AWS storage distribution algorithm
```
- Change the existing implementation of SortByIOPS and FilterByIOPS to consider
  an IOPS range.
- In case of AWS, the size of the disk dictates the IOPS. Moved the logic of
  determining the IOPS of a StoragePool from the common algorithm to per cloud based logic.
- ResizeDisk: For resize disk instead of considering only one row which fits the current drive
  size and iops. Choose the all the rows which allow the existing IOPS and the drive size.
```

### CODE REVIEW

File: vsphere/storagemanager/vsphere.go [hunk 1]
```diff
@@ -23,12 +23,37 @@ func newVsphereStorageManager(
 func (a *vsphereStorageManager) GetStorageDistribution(
 	request *cloudops.StorageDistributionRequest,
 ) (*cloudops.StorageDistributionResponse, error) {
-	return storagedistribution.GetStorageDistribution(request, a.decisionMatrix)
+	response := &cloudops.StorageDistributionResponse{}
+	for _, userRequest := range request.UserStorageSpec {
+		// for for request, find how many instances per zone needs to have storage
+		// and the storage spec for each of them
+		instStorage, instancePerZone, _, err :=
```
Comment (harsh-px) [inline vsphere/storagemanager/vsphere.go]: nit: rename to instancesPerZone

File: specs/decisionmatrix/aws.yaml [hunk 2]
```diff
@@ -0,0 +1,3620 @@
+rows:
```
Comment (harsh-px) [inline specs/decisionmatrix/aws.yaml]: in the generated file, add a comment to the top that this file is generated automatically using the the make rule and do not edit manually

File: aws/storagemanager/spec/Makefile [hunk 3]
```diff
@@ -0,0 +1,4 @@
+gen:
```
Comment (harsh-px) [inline aws/storagemanager/spec/Makefile]: we should probably run the gen rule in travis and fail if the aws.yaml generated is different from the one checked in. Can be in a followup PR.

File: aws/storagemanager/aws_test.go [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+						DriveCapacityGiB: 700,
+						DriveType:        "gp2",
+						DriveCount:       1,
+						IOPS:             2100,
+					},
+				},
+			},
+			expectedErr: nil,
+		},
+		/*{
```
Comment (harsh-px) [inline aws/storagemanager/aws_test.go]: is this commented deliberately?

File: cloud_storage_management.go [hunk 5]
```diff
@@ -208,23 +210,41 @@ func (dm *StorageDecisionMatrix) FilterByDriveType(requestedDriveType string) *S
 	return dm
 }
 
-// FilterByIOPS filters out the rows which do not meet the requested IOPS.
+// FilterByMinIOPS filters out the rows whose minIOPS are greater requested IOPS.
```
Comment (harsh-px) [inline cloud_storage_management.go]: change to `filters out the rows whose minIOPS are less than requested IOPS'

File: cloud_storage_management.go [hunk 6]
```diff
@@ -237,6 +257,23 @@ func (dm *StorageDecisionMatrix) FilterByDriveSize(currentDriveSize uint64) *Sto
 	return dm
 }
 
+// FilterByDriveSize filters out the rows for which the current drive size is greater than
+// row's min size
```
Comment (harsh-px) [inline cloud_storage_management.go]: add function comment to cover the `else if`

File: pkg/storagedistribution/storagedistribution.go [hunk 7]
```diff
@@ -348,9 +312,10 @@ row_loop:
 
 	if int(rowIndex) == len(dm.Rows) {
 		// row_loop failed
-		return nil, 0, cloudops.ErrStorageDistributionCandidateNotFound
+		return nil, 0, nil, cloudops.ErrStorageDistributionCandidateNotFound
 	}
 
+	fmt.Println("$$$$ minCapacityPerZone: ", minCapacityPerZone, driveCount, driveSize)
```
Comment (harsh-px) [inline pkg/storagedistribution/storagedistribution.go]: remove

File: aws/storagemanager/aws_test.go (hunk 4 above)
Comment (adityadani) [reply aws/storagemanager/aws_test.go]: this is the same test case which has been commented for all providers. The algorithm currently cannot handle adding disks to storage less nodes.

File: aws/storagemanager/spec/Makefile (hunk 3 above)
Comment (adityadani) [reply aws/storagemanager/spec/Makefile]: good idea. Will do.

File: pkg/storagedistribution/storagedistribution.go (hunk 7 above)
Comment (adityadani) [reply pkg/storagedistribution/storagedistribution.go]: ah! this one got slipped in. Will fix

**Pull request merged**